### PR TITLE
improved initialization for large towns

### LINKED ISSUE #32: Simulation starts before initialization of all the actors
For large towns like 600x400, first node starts simulation before initialization of its actors' neighbours on the last node which causes non-delivered messages and simulation blocks on first time steps.

### PR DESCRIPTION
**Pull request by mjanczykowski**
All actors resolve their neighbors' actorRefs from paths before starting simulation. Fixes #32 
Using resolveOne to check if actors are alive. And got rid of workers unregistration - now they watch for disassociation events.

### CODE REVIEW

File: common/src/main/scala/miss/supervisor/Supervisor.scala [hunk 1]
```diff
@@ -226,9 +223,24 @@ class Supervisor(config: Config) extends FSM[State, Data] {
         actors(i)(j) ! StartSimulation(areaVerticalRoadDefs.toList, areaHorizontalRoadDefs.toList, i, j)
       }
     }
+    log.info("Done sending StartSimulation")
     actors
   }
 
+  private def handleRemotingEvents: StateFunction = {
+    case Event(ev@DisassociatedEvent(_, _, _), SupervisorData(workers, actors, _)) =>
```
Comment (salceson) [inline common/src/main/scala/miss/supervisor/Supervisor.scala]: ```case Event(ev: DisassociatedEvent, SupervisorDate(workers, actors, _))```

File: common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala [hunk 2]
```diff
@@ -1,34 +1,104 @@
...
+  }
+
+  private def resolveActor(actorPath: ActorPath): Unit = {
+    log.debug(s"Resolving $actorPath")
+
+    def resolveActorWithRetries(actorPath: ActorPath, retries: Int)(implicit ec: ExecutionContext): Future[ActorRef] = {
+      val future = this.context.actorSelection(actorPath).resolveOne(resolveActorTimeoutSeconds seconds)
+      future onComplete {
+        case Failure(e) if retries > 0 =>
+          log.info(s"Cannot resolve actor: $actorPath due to: ${e.getMessage}. Retrying.")
```
Comment (salceson) [inline common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala]: Maybe log with retry time? I mean `retrying for 2nd time` or something similar.

File: common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala [hunk 3]
```diff
@@ -1,34 +1,104 @@
...
   startWith(Initialized, EmptyData)
 
+  private def retry[T](f: => Future[T], delay: FiniteDuration, retries: Int)(implicit ec: ExecutionContext): Future[T] = {
+    f recoverWith { case _ if retries > 0 => after(delay, context.system.scheduler)(retry(f, delay, retries - 1)) }
+  }
+
+  private def resolveActor(actorPath: ActorPath): Unit = {
+    log.debug(s"Resolving $actorPath")
+
+    def resolveActorWithRetries(actorPath: ActorPath, retries: Int)(implicit ec: ExecutionContext): Future[ActorRef] = {
```
Comment (salceson) [inline common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala]: Why are you not using here `retry[T]`?

Comment (salceson) [reply common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala]: Also `actorPath` parameter is not needed IMO. It can be read from the closure.

File: common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala [hunk 4]
```diff
@@ -1,34 +1,104 @@
...
+    }
+
+    resolveActorWithRetries(actorPath, resolveActorRetries)
+  }
+
   when(Initialized) {
-    case Event(StartSimulation(verticalRoadsDefs, horizontalRoadsDefs, x, y), EmptyData) =>
+    case Event(msg@StartSimulation(verticalRoadsDefs, horizontalRoadsDefs, x, y), EmptyData) =>
+      val actorsToResolve = verticalRoadsDefs.flatMap(vrd => Set(vrd.prevAreaActorPath, vrd.outgoingActorPath)).toSet ++ horizontalRoadsDefs.flatMap(hrd => Set(hrd.prevAreaActorPath, hrd.outgoingActorPath)).toSet
+      //      actorsToResolve.foreach(actorPath => resolveActor(actorPath))
```
Comment (salceson) [inline common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala]: Why is it commented?

File: common/src/main/scala/miss/trafficsimulation/roads/Road.scala [hunk 5]
```diff
@@ -179,7 +179,7 @@ class RoadSegment(val roadId: RoadId,
     }
 
     //switch lanes not supported on initial part of firstRoadSegment
-    if(1==1) {
+    if (1 == 1) {
```
Comment (salceson) [inline common/src/main/scala/miss/trafficsimulation/roads/Road.scala:182]: Wtf? Condition always true.

**Review by salceson — CHANGES_REQUESTED** [review]
Can you summarize in "high level" what changes did you do here?

Also found out some things that I think needs changing. Please look at my comments :)

File: common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala (hunk 3 above)
Comment (mjanczykowski) [reply common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala]: Removed retry method. I don't know if futures are reusable, but it wasn't working. Found it on some scala forums.

File: common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala (hunk 4 above)
Comment (mjanczykowski) [reply common/src/main/scala/miss/trafficsimulation/actors/AreaActor.scala]: Removed. I was first trying to resolve all actors at once but it caused dead locks when there were thousands of futures for opposite resolveOne requests.

File: common/src/main/scala/miss/trafficsimulation/roads/Road.scala (hunk 5 above)
Comment (mjanczykowski) [reply common/src/main/scala/miss/trafficsimulation/roads/Road.scala:182]: no idea, only reformatted the file :) I'll remove it

Comment (mjanczykowski) [reply common/src/main/scala/miss/trafficsimulation/roads/Road.scala:182]: maybe there should be some variable used? (looking at the comment above)

Comment (mjanczykowski) [reply common/src/main/scala/miss/trafficsimulation/roads/Road.scala:182]: finally, I'll leave it :)

Comment (salceson) [reply common/src/main/scala/miss/trafficsimulation/roads/Road.scala:182]: Yeah, I understand that it's reformated only. But this is **very** suspicious. I'll look at this today - a bit later.
